colcon_core/shell/__init__.py

```diff
@@ -310,3 +310,40 @@ def create_environment_hook(
             'Could not find a primary shell extension for creating an '
             'environment hook')
     return hooks
+
+
+_colcon_prefix_path_cache = {}
```

Comment:
Why cache the result? Can the `COLCON_PREFIX_PATH` environment variable change between calls to `get_colcon_prefix_path()`?

Reply:
I tried to describe that in the docblock:

> The result is being cached based on the passed skip parameter to avoid printing the warning on repeated invocations.

While the environment isn't supposed to change during one run of the tool the fact that the result would be invalid if the environment does change isn't really a good idea.

I refactored the logic in 33b1d05e9ced08f109ad31ecb13ad1c8b79b50c9 to actually store which warning has already been logged and just use that global information to avoid repeating the warning. The computation happens every time and is therefore correct no matter what happens to the environment.